mlflow/pyspark/ml/__init__.py

```diff
@@ -107,24 +120,128 @@ def _get_estimator_info_tags(estimator):
+    def gen_instance_name(instance):
+        stage_cls = instance.__class__.__name__
+        stage_uid = instance.uid
 
-    stage_hierarchy = []
-    pipeline_stages = pipeline.getStages()
-    for stage in pipeline_stages:
-        if isinstance(stage, Pipeline):
-            hierarchy_elem = _get_pipeline_stage_hierarchy(stage)
+        if stage_cls not in cls2count_map:
+            cls2count_map[stage_cls] = 1
+            uid2name_map[stage_uid] = stage_cls
+            cls2first_uid_map[stage_cls] = stage_uid
         else:
-            hierarchy_elem = stage.uid
-        stage_hierarchy.append(hierarchy_elem)
-    return {pipeline.uid: stage_hierarchy}
+            cls2count_map[stage_cls] += 1
+            uid2name_map[stage_uid] = f'{stage_cls}_{cls2count_map[stage_cls]}'
+
+        return uid2name_map[stage_uid]
+
+    instance_name = gen_instance_name(instance)
+    if isinstance(instance, Pipeline):
+        sub_stages = []
+        for sub_stage in instance.getStages():
+            sub_stages.append(_gen_instance_metadata_recusively(
+                sub_stage, uid2name_map, cls2count_map, cls2first_uid_map, param_search_estimators
+            ))
+        return {'name': instance_name, 'stages': sub_stages}
+    elif isinstance(instance, OneVsRest):
+        classifier = _gen_instance_metadata_recusively(
+            instance, uid2name_map, cls2count_map, cls2first_uid_map
+        )
+        return {'name': instance_name, 'classifier': classifier}
+    elif _is_parameter_search_estimator(instance):
+        param_search_estimators.append(instance)
+        evaluator = instance.getEvaluator()
+        tuned_estimator = instance.getEstimator()
+        return {
+            'name': instance_name,
+            'evaluator': _gen_instance_metadata_recusively(
+                evaluator, uid2name_map, cls2count_map, cls2first_uid_map, param_search_estimators
+            ),
+            'tuned_estimator': _gen_instance_metadata_recusively(
+                tuned_estimator, uid2name_map, cls2count_map, cls2first_uid_map,
+                param_search_estimators
+            )
+        }
+    else:
+        return instance_name
 
 
-def _get_instance_param_map_recursively(instance, level):
+def _gen_estimator_metadata(estimator):
+    """
+    Returns an AutologgingEstimatorMetadata object.
+    The AutologgingEstimatorMetadata object includes:
+     - hierarchy: the hierarchy of the estimator, it will expand
+         pipeline/meta estimator/param tuning estimator
+     - uid2name_map: a map of `uid` -> `name`, each nested instance uid will be
+         mapped to a fixed name. The naming rule is using `{class_name}` if the
+         instance type occurs once, or `{class_name}_{index}` if the instance type occurs
+         multiple times. The index order is in line with depth-first traversing.
+     - param_search_estimators: a list includes all param search estimators in the
+         hierarchy tree.
+    """
+    uid2name_map = {}
```

Comment:
Can we just generate `uid2name_map` first, then generate other matadata using it to simplify the code?

```python
uid2name_map = ...

hierach = func(..., uid2name_map)
...
```